### ISSUE TITLE
'and' and 'or' operator suport for xpath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2502]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jack reported|
|-|-|
|Trac ID|trac#2502|
|Type|enhancement|
|Created|2007-03-02 17:13:30Z|

The attached patch (with updated and addition test cases) adds 'and' and 'or' operators to the xpath grammar.  Now xpath predicate expressions can be more complex (example: /iq[@from='someone@chesspark.com/cpc' and (@type='set' or @type='get')]).  

This makes programming with twisted words much easier as callback functions can be made very specific, instead of having to have routing functions doing the extra work.

Also, writing test cases is significantly easier, since you can use the xpath matcher to do much of the error checking work.

The patch is very simple, but there are lots of minor doc edits because someone had not been keeping the grammar file updated (they were just changing the generated output).

Attachments:

* [xpath.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/249/2491991e0c2b3ea58ff0f92851ec4af9133a2ca1/efc5fdedf28579ecbb1e87dc1a4d58e7d24be135.diff) (22679 bytes) - added by jack on 2007-03-02 21:19:40Z - new operators for xpath

<details><summary>Searchable metadata</summary>

```
trac-id__2502 2502
type__enhancement enhancement
reporter__jack jack
priority__highest highest
milestone__Words_0_6 Words-0.6
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1172855610000000 1172855610000000
changetime__1184160713000000 1184160713000000
version__None None
owner__ 
cc__twonds cc__ralphm cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jack's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|jack commented|
|-|-|

Oops.  There was a small bug in the grammar that made three operands not work.  I've modified the patch to test for this and fixed the problem.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)|@ralphm set owner to @ralphm|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_3"></a>|@ralphm removed owner|
|-|-|

I created a branch [branches/xish-xpathparser-2502](branches/xish-xpathparser-2502) and imported Jack's patch there. I reviewed the actual changes and tests and they look fine. Thanks!

Because I wasn't happy with how `xpathparser.py` was generated, I included the Yapps runtime in a separate module, added comments on how the generated code file should not be edited and did some general cleanup.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to @ralphm|
|-|-|

The granularity of test_staticMethods in test_xpath.py doesn't seem right.  The "static methods" of xpath aren't a unit - probably xpath.matches is at least one, xpath.queryForNodes is at least one, xpath.queryForString is at least one, and xpath.queryForStringList is at least one.  The vagueness of the docstring somewhat reflects this - if the test had more direction, it would probably be easier to describe.  Other tests in this module have similar problems.  The size of the document on which they're all operating also suggests they're doing too much - few of the tests need more than 3 or 4 nodes to cover the behavior they're trying to cover, but they're all operating on a document with more than 20 nodes.  This is really a pre-existing issue with the code, but I'm pointing it out since a lot of tests have minor changes made to them in this branch, and some new tests are added in the same style.

Thanks for making the status of xpathparser.py more explicit :)  One change I'd suggest is making the yappsrt import absolute.  That is, ```from twisted.words.xish import yappsrt as runtime```.

In xpath.py, BooleanValue should document its instance variables and _booleanAnd and _booleanOr should probably have docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_5"></a>|@ralphm set owner to @exarkun|
|-|-|

Yes those tests do not deserve any price for aesthetics. That should be changed in the future. Deferring that now.

As for the other comments, those have been addressed in the branch. I suppose an Interface description would be a good idea for better documentation down the line.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @ralphm|
|-|-|

Looks good, go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set status to `closed`|
|-|-|

(In [20496]) Add `and` and `or` operators for Xish XPath expressions.

Author: jack (ralphm)
Reviewer: ralphm, exarkun
Fixes [#2502](https://github.com/twisted/twisted/issues/2502).

This change also uses the separate yapps runtime, simplifying future changes
to the grammar.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `reopened`|
|-|-|

Reverted at r20528:

  * doc buildslave has failures - let's fix these by pulling yappsrt out of Twisted completely, I think that'll resolve the issue
  * the generated xpathparser.py also has manual modifications, which we should avoid :)

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_9"></a>|@ralphm commented|
|-|-|

For reference, this failure on the doc buildslave is covered in [#2701](https://github.com/twisted/twisted/issues/2701).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

For the record, the code for this ticket is still in xish-xpathparser-2502.  The yapps code has been moved out of twisted/words/.  If we want to distribute it, then it seems like the thing to do is to put it into twisted/words/topfiles and adjust the setup.py to install it.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_11"></a>|@ralphm set owner to @exarkun<br>@ralphm set status to `new`|
|-|-|

Modified [branches/xish-xpathparser-2502-2](branches/xish-xpathparser-2502-2) two weeks ago to reembed the yapps runtime and better document the generation process.

Please review again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @ralphm|
|-|-|

The sed expression seems to be slightly buggy ;)

```
# 3.) Edit the output to depend on the embedded runtime, not 
```

There's also some trailing whitespace in xpathparser.g, which I suppose should be cleaned up.

The rest seems kosher.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_13"></a>|@ralphm commented|
|-|-|

Heh. Oops. Fixed and committed. Merging.

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_14"></a>|@ralphm set status to `closed`|
|-|-|

(In [20726]) Add and and or operators for Xish XPath expressions.

Author: jack (ralphm)
Reviewer: ralphm, exarkun
Fixes [#2502](https://github.com/twisted/twisted/issues/2502).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
